blog-post/blog-post.md

```diff
@@ -0,0 +1,634 @@
+
+We think that taken together these core primitives will enable developers to produce stronger
+abstractions, write more modular and reusable code, leverage the type system to enforce core
+safety properties.
+
+We will continue to support the kind of low level access to the EVM that is often required by
+production implementations: assembly will remain a core primitive with an important new feature:
+we will support calling functions from the high level language in assembly directly. Users will
+be able to disable the built in abstractions (e.g. contract dispatch generation, ABI decoding,
+default storage layout generation), following the "pay for what you use" philosophy of languages
+like Rust and C++.
+
+Aside from the removal of inheritance, we expect to be able to support the majority of existing
+language features without breakage.
+
+### Generics and type classes
+
+Core Solidity introduces two new exciting abstraction mechanisms: generics and
+type classes.
+
+Generics enable parametric polymorphism through type parameters, which allows the
+implementation of algorithms and data structures that operate uniformly across
+all types. As an example, we could define a polymorphic `identity` function:
+
+```
+forall T . function identity(x : T) -> T {
+    return x;
+}
+```
+
+While generic functions are interesting, most interesting operations are not defined
+for all types. Overloading allows the definition of code which can operate in distinct
+ways for different types. Type classes are the standard way of combining overloading and
+parametric polymorphism (generics) in a systematic manner. Type classes are similar to
+Rust traits: they provide signatures for the functions which will be implemented, for
+distinct types, in instance definitions. In this sense, instance declarations are similar
+to `impl` in Rust.
+
+A type class definition declares the class name, its arguments and member functions type
+signatures. As example, let's consider the following definition of a class for the ABI
+encoding of types:
+
+```
+forall self . class self:ABIEncode {
+    function encodeInto(x:self, basePtr:word, offset:word, tail:word) -> word;
+}
+```
+
+Function `encodeInto` encodes a value of type `self` into a memory region starting at
+`basePtr`. Argument `offset` denotes the position of the first empty byte of the head
+and `tail` which gives the position of the first empty byte of the tail. The
+implementation of encoding for a specific type is done in an instance declaration.
+As an example, the following instance definition implements the ABI encoding for
+`uint256` type, which is represented as an algebraic type that holds a `word` value.
+
+```
+data uint256 = uint256(word);
+instance uint256:ABIEncode {
+    function encodeInto(x:uint256, basePtr:word, offset:word, tail:word) -> word {
+        let repx : word = Typedef.rep(x);
+        assembly { mstore(add(basePtr, offset), repx) }
+        return tail;
+    }
+}
+```
```

Comment:
- This explains parameters, but does not say that the return value is (and it is unnamed).
- Both `offset` and `tail` are relative positions, right? I'd say that explicitly.
- Wouldn't `head` be a better name for `offset`? In the context of encoding I usually associate "offset" with the tail offset, so naming here is not really intuitive.